From epasheva@sierrawireless.com  Tue May 12 17:13:35 2009
From: Elina Pasheva <epasheva@sierrawireless.com>
Date: Tue, 12 May 2009 13:12:24 -0700
Subject: USB: serial: sierra driver zero length packet fix
To: <gregkh@suse.de>
Cc: <rfiler@sierrawireless.com>, <epasheva@sierrawireless.com>
Message-ID: <1242159144.7032.15.camel@Linuxdev3>

From: Elina Pasheva <epasheva@sierrawireless.com>

- Fixed a problem with transferring packets with size a multiple of Bulk
  Xfer size in function sierra_write(). Added transfer flag
  URB_ZERO_PACKET before submitting the urb to trigger Zero-length data
  transfer when packet size is a multiple of Bulk Xfer.

Signed-off-by: Elina Pasheva <epasheva@sierrawireless.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>


---
 drivers/usb/serial/sierra.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/drivers/usb/serial/sierra.c
+++ b/drivers/usb/serial/sierra.c
@@ -17,7 +17,7 @@
   Whom based his on the Keyspan driver by Hugh Blemings <hugh@blemings.org>
 */
 
-#define DRIVER_VERSION "v.1.3.6"
+#define DRIVER_VERSION "v.1.3.7"
 #define DRIVER_AUTHOR "Kevin Lloyd, Elina Pasheva, Matthew Safar, Rory Filer"
 #define DRIVER_DESC "USB Driver for Sierra Wireless USB modems"
 
@@ -428,6 +428,9 @@ static int sierra_write(struct tty_struc
 					  port->bulk_out_endpointAddress),
 			  buffer, writesize, sierra_outdat_callback, port);
 
+	/* Handle the need to send a zero length packet */
+	urb->transfer_flags |= URB_ZERO_PACKET;
+
 	/* send it down the pipe */
 	retval = usb_submit_urb(urb, GFP_ATOMIC);
 	if (retval) {
